### ISSUE TITLE
Dots in version numbers are causing zero version numbers

### ISSUE THREAD

**Issue by delphidabbler** [delphidabbler/vied#55]
If a macro uses a version number with dots in it then the version number is being parsed as `0.0.0.0`.

This bug seems to have been introduced in hot fix v2.14.1

**Comment by delphidabbler** [comment]
It's true the bug isn't in v2.14.0 and appears in v2.14.1, but I can't for the life of me see why.

v2.14.1 changed nothing to do with the bug! Here's the diff between the two:

```
diff --git a/CHANGELOG.md b/CHANGELOG.md
index 1576ea8..e73cb14 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 # Change Log for Version Information Editor
 
+## v2.14.1 of 11 April 2023
+
++ HOTFIX of potential buffer overrun in code that copies resource file statements to clipboard [issue #54].
+
 ## v2.14.0 of 28 August 2022
 
 + New feature that enables the user to define "macros" that can be used to insert text in one or more version information strings and in the file and product version fixed file information entries. There are three types of macro:
diff --git a/Source/UVInfo.pas b/Source/UVInfo.pas
index bb10fcd..34ba11a 100644
--- a/Source/UVInfo.pas
+++ b/Source/UVInfo.pas
@@ -614,40 +614,12 @@ procedure TVInfo.CopyToClipBoard;
   }
 var
   RCList: TStringList;          // string list to hold resource file code
-  I: Integer;                   // loop control
-  PBuf: PChar;                  // buffer to hold whole  text to copy
-  BufSize: Integer;             // size of above buffer
-  Line: array[0..257] of char;  // text of each line of resource file
 begin
-  // Give PBuf a nil value so that StrDispose will always work
-  PBuf := nil;
-  // Create the string list
   RCList := TStringList.Create;
   try
-    // Write text of "file" to string list
     WriteAsRC(RCList);
-    // Calcualte size required for buffer to hold all text
-    // size 1 for terminal \0 character
-    BufSize := 1;
-    // itereate through lines incrementing by length of line (+ 2 for EOL)
-    for I := 0 to RCList.Count - 1 do
-      Inc(BufSize, Length(RCList[I]) + 2);
-    // Allocate buffer of required size and set to empty string
-    PBuf := StrAlloc(BufSize);
-    StrCopy(PBuf, '');
-    // Iterate through list of lines, appending each line to the buffer
-    for I := 0 to RCList.Count - 1 do
-    begin
-      // copy line + CR + LF to C format string
-      StrPCopy(Line, RCList[I] + #13#10);
-      // append line to end of buffer
-      StrCat(PBuf, Line);
-    end;
-    // Copy whole buffer to clipboard
-    ClipBoard.SetTextBuf(PBuf);
+    Clipboard.AsText := TrimRight(RCList.Text) + #13#10;
   finally
-    // Free buffer and string list
-    StrDispose(PBuf);
     RCList.Free;
   end;
 end;
diff --git a/Source/VERSION b/Source/VERSION
index edcfe40..b70ae75 100644
--- a/Source/VERSION
+++ b/Source/VERSION
@@ -1 +1 @@
-2.14.0
+2.14.1
```

There nothing there that I can see that caused this. 🤯

**Comment by delphidabbler** [comment]
Found the culprit:

```pascal
function StrToVersionNumber(const Str: string): TPJVersionNumber;
  {Converts string in format used in VERSIONINFO resource statement to
  TPJVersionNumber type.
    @param Str [in] String representation of version number.
    @return Required TPJVersion information structure.
  }
var
  Start: Integer;               // cursor to walk along string}
  NumStr: string;               // current number in string}
  Nums: array[1..4] of LongInt; // array to hold numbers from string}
  I: Integer;                   // loop control variable}
begin
  Start := 1;
  // Iterate for each version number (there are four in string)
  for I := Low(Nums) to High(Nums) do
  begin
    NextField(Str, Start, NumStr, ',');
    NumStr := TrimSpaces(NumStr);
    // Convert NumStr into the required number - default 0 if no valid value
    if NumStr <> '' then
      Nums[I] := StrToIntDef(NumStr, 0)
    else
      Nums[I] := 0;
  end;
  Result.V1 := Nums[1];
  Result.V2 := Nums[2];
  Result.V3 := Nums[3];
  Result.V4 := Nums[4];
end;
```

This code assumes that version number are comma separated, so never splits the version number into fields and therefore parses only one field that contains more than 1 dot -> StrToIntDef returns 0.

This code didn't change from v2.14.0 to v2.14.1 so huh? 🤯 

My best guess is that v2.14.0 got compiled with some code changes that never got committed, which is deeply annoying!

**Comment by delphidabbler** [comment]
Fixed in HOTFIX release v2.14.2 and merge commit ef8b65e1d28fb56bce0aee60a454df0858c7f06d
